### PR TITLE
8330844: Add aliases for conditional jumps and additional instruction forms for x86

### PR DESCRIPTION
**Pull request by asgibbons**
Adding infrastructure for JDK-8320448.  Aliasing conditional jump instructions; adding some x86 instructions.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8330844](https://bugs.openjdk.org/browse/JDK-8330844): Add aliases for conditional jumps and additional instruction forms for x86 (**Enhancement** - P4)

### Reviewers
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**) ⚠️ Review applies to [f7d7f7de](https://git.openjdk.org/jdk/pull/18893/files/f7d7f7de2c192f3f73eac36ca8114c937b8671a0)
 * [Emanuel Peter](https://openjdk.org/census#epeter) (@eme64 - **Reviewer**) ⚠️ Review applies to [fe5f3060](https://git.openjdk.org/jdk/pull/18893/files/fe5f3060cc093683a76ffcb9916c8e0313edc1ab)
 * [Sandhya Viswanathan](https://openjdk.org/census#sviswanathan) (@sviswa7 - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/18893/head:pull/18893` \
`$ git checkout pull/18893`

Update a local copy of the PR: \
`$ git checkout pull/18893` \
`$ git pull https://git.openjdk.org/jdk.git pull/18893/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 18893`

View PR using the GUI difftool: \
`$ git pr show -t 18893`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/18893.diff">https://git.openjdk.org/jdk/pull/18893.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/18893#issuecomment-2070126996)

### CODE REVIEW

**Comment by asgibbons** [comment]
This is a precursor for [JDK-8320448](https://bugs.openjdk.org/browse/JDK-8320448), essentially adding infrastructure requirements for that algorithm.

**Comment by vnkozlov** [comment]
> making arrays_equals accessible from stubs

I am not sure I understand why you need to move it.  Your changes for JDK-8320448 shows that new code is used only by C2. You can move your new code in stubGenerator_x86_64.cpp into the part under `#ifdef COMPILER2`. And code in `stubGenerator_x86_64_string.cpp` could be put under this `#ifdef` too.

**Comment by asgibbons** [comment]
Adding the `#ifdef COMPILER2` in `stubGenerator_x86_64_string.cpp` allows for good compilation for JDK-8320448, so I can undo the move.  Thanks for spotting that.

**Comment by vnkozlov** [comment]
Can you also remove changes in `arrays_equals` from this PR? It is fine to have them in JDK-8320448 changes.

File: src/hotspot/cpu/x86/c2_MacroAssembler_x86.cpp [hunk 1]
```diff
@@ -4489,12 +4489,19 @@ void C2_MacroAssembler::count_positives(Register ary1, Register len,
...
+                                      Register result, Register chr,
+                                      XMMRegister vec1, XMMRegister vec2,
+                                      bool is_char, KRegister mask,
+                                      bool expand_ary2) {
+  // for expand_ary2, limit is the (smaller) size of the second array.
   ShortBranchVerifier sbv(this);
   Label TRUE_LABEL, FALSE_LABEL, DONE, COMPARE_VECTORS, COMPARE_CHAR, COMPARE_BYTE;
 
+  assert((!expand_ary2) || ((expand_ary2) && (UseAVX == 2)),
+         "Expansion only implemented for AVX2");
```
Comment (vnkozlov) [inline src/hotspot/cpu/x86/c2_MacroAssembler_x86.cpp:4503]: BTW, the check in assert could be simplified: `(!expand_ary2 || UseAVX == 2)`

Comment (asgibbons) [reply src/hotspot/cpu/x86/c2_MacroAssembler_x86.cpp:4503]: I thought this would make the intent explicitly clear.

**Comment by asgibbons** [comment]
A large part of this PR was to lessen the burden of reviewing JDK-8320448 changes.  Am I hearing you say that this approach is not desired?  The other PR is a big review and I was hoping to piecemeal some non-core algorithm changes in to make the review easier.

It is, of course, trivial to revert the change to arrays_equals.  Please let me know your final decision.  Thanks.

**Comment by vnkozlov** [comment]
> A large part of this PR was to lessen the burden of reviewing JDK-8320448 changes. Am I hearing you say that this approach is not desired? The other PR is a big review and I was hoping to piecemeal some non-core algorithm changes in to make the review easier.
> 
> It is, of course, trivial to revert the change to arrays_equals. Please let me know your final decision. Thanks.

I am for splitting big PRs if possible. And you are not limited how many self-containing sub-PRs you can create.
But each PR should address one issue for easy review and testing.

I consider this PR should address what in its title: aliases for jump instructions and adding missing cmp/jump instructions (which is related).

Any changes to not related code, like arrays_equals, do not belong here. It could be separate sub-PR or even followup PR.

**Comment by asgibbons** [comment]
OK.  arrays_equals changes reverted.

**Review by vnkozlov — APPROVED** [review]
Good.

**Comment by asgibbons** [comment]
/integrate

Thanks!

File: src/hotspot/cpu/x86/macroAssembler_x86.hpp [hunk 2]
```diff
@@ -955,6 +955,74 @@ class MacroAssembler: public Assembler {
...
+  void ALWAYSINLINE jg(Label& L, bool maybe_short = true) { jcc(Assembler::greater, L, maybe_short); }
+  void ALWAYSINLINE jnle(Label& L, bool maybe_short = true) { jcc(Assembler::greater, L, maybe_short); }
+  void ALWAYSINLINE jp(Label& L, bool maybe_short = true) { jcc(Assembler::parity, L, maybe_short); }
+  void ALWAYSINLINE jpe(Label& L, bool maybe_short = true) { jcc(Assembler::parity, L, maybe_short); }
+  void ALWAYSINLINE jnp(Label& L, bool maybe_short = true) { jcc(Assembler::noParity, L, maybe_short); }
+  void ALWAYSINLINE jpo(Label& L, bool maybe_short = true) { jcc(Assembler::noParity, L, maybe_short); }
+// * No condition for this *  void ALWAYSINLINE jcxz(Label& L, bool maybe_short = true) { jcc(Assembler::cxz, L, maybe_short); }
+// * No condition for this *  void ALWAYSINLINE jecxz(Label& L, bool maybe_short = true) { jcc(Assembler::cxz, L, maybe_short); }
+
+// Short versions of the above
```
Comment (eme64) [inline src/hotspot/cpu/x86/macroAssembler_x86.hpp:992]:
```suggestion
  // * No condition for this *  void ALWAYSINLINE jcxz(Label& L, bool maybe_short = true) { jcc(Assembler::cxz, L, maybe_short); }
  // * No condition for this *  void ALWAYSINLINE jecxz(Label& L, bool maybe_short = true) { jcc(Assembler::cxz, L, maybe_short); }

  // Short versions of the above
```

File: src/hotspot/cpu/x86/macroAssembler_x86.hpp [hunk 3]
```diff
@@ -955,6 +955,74 @@ class MacroAssembler: public Assembler {
...
+  void ALWAYSINLINE jle_b(Label& L) { jccb(Assembler::lessEqual, L); }
+  void ALWAYSINLINE jng_b(Label& L) { jccb(Assembler::lessEqual, L); }
+  void ALWAYSINLINE jg_b(Label& L) { jccb(Assembler::greater, L); }
+  void ALWAYSINLINE jnle_b(Label& L) { jccb(Assembler::greater, L); }
+  void ALWAYSINLINE jp_b(Label& L) { jccb(Assembler::parity, L); }
+  void ALWAYSINLINE jpe_b(Label& L) { jccb(Assembler::parity, L); }
+  void ALWAYSINLINE jnp_b(Label& L) { jccb(Assembler::noParity, L); }
+  void ALWAYSINLINE jpo_b(Label& L) { jccb(Assembler::noParity, L); }
+// * No condition for this *  void ALWAYSINLINE jcxz_b(Label& L) { jccb(Assembler::cxz, L); }
+// * No condition for this *  void ALWAYSINLINE jecxz_b(Label& L) { jccb(Assembler::cxz, L); }
```
Comment (eme64) [inline src/hotspot/cpu/x86/macroAssembler_x86.hpp:1024]:
```suggestion
  // * No condition for this *  void ALWAYSINLINE jcxz_b(Label& L) { jccb(Assembler::cxz, L); }
  // * No condition for this *  void ALWAYSINLINE jecxz_b(Label& L) { jccb(Assembler::cxz, L); }
```

**Review by eme64 — APPROVED** [review]
Looks reasonable. Can you apply the indentation issue, please?

File: src/hotspot/cpu/x86/macroAssembler_x86.hpp (hunk 3 above)
Comment (asgibbons) [reply src/hotspot/cpu/x86/macroAssembler_x86.hpp]: Done

**Comment by eme64** [comment]
Thanks for the update. I can sponsor as soon as you attempt integration again.

**Comment by asgibbons** [comment]
Thank you.  I'm waiting on @sviswa7 review before integrating again

File: src/hotspot/cpu/x86/macroAssembler_x86.hpp [hunk 4]
```diff
@@ -955,6 +955,74 @@ class MacroAssembler: public Assembler {
   // contained in the location described by entry (not the address of entry)
   void jump(ArrayAddress entry, Register rscratch);
 
+  // Adding more natural conditional jump instructions
+  void ALWAYSINLINE jo(Label& L, bool maybe_short = true) { jcc(Assembler::overflow, L, maybe_short); }
+  void ALWAYSINLINE jno(Label& L, bool maybe_short = true) { jcc(Assembler::noOverflow, L, maybe_short); }
+  void ALWAYSINLINE js(Label& L, bool maybe_short = true) { jcc(Assembler::positive, L, maybe_short); }
```
Comment (sviswa7) [inline src/hotspot/cpu/x86/macroAssembler_x86.hpp:961]: Isn't js -> jump is sign flag is set -> Assembler::negative?
Correspondingly jns, js_b, jns_b should also be corrected.

File: src/hotspot/cpu/x86/assembler_x86.cpp [hunk 5]
```diff
@@ -1804,6 +1827,14 @@ void Assembler::cmpw(Address dst, int imm16) {
   emit_int16(imm16);
 }
 
+void Assembler::cmpw(Address dst, Register reg) {
+  InstructionMark im(this);
+  emit_int8(0x66);
+  prefix(dst, reg);
+  emit_int8((unsigned char)0x39);
+  emit_operand(reg, dst, 1);
```
Comment (sviswa7) [inline src/hotspot/cpu/x86/assembler_x86.cpp:1835]: This should be emit_operand(reg, dst, 0);

File: src/hotspot/cpu/x86/assembler_x86.cpp [hunk 6]
```diff
@@ -4531,6 +4571,16 @@ void Assembler::vpcmpeqw(XMMRegister dst, XMMRegister nds, XMMRegister src, int
   emit_int16(0x75, (0xC0 | encode));
 }
 
+// In this context, the dst vector contains the components that are equal, non equal components are zeroed in dst
+void Assembler::vpcmpeqw(XMMRegister dst, XMMRegister nds, Address src, int vector_len) {
+  assert(vector_len == AVX_128bit ? VM_Version::supports_avx() : VM_Version::supports_avx2(), "");
```
Comment (sviswa7) [inline src/hotspot/cpu/x86/assembler_x86.cpp:4576]: InstructionMark missing in this instruction which takes Address as operand?

File: src/hotspot/cpu/x86/assembler_x86.cpp [hunk 7]
```diff
@@ -4425,6 +4456,15 @@ void Assembler::vpcmpeqb(XMMRegister dst, XMMRegister nds, XMMRegister src, int
   emit_int16(0x74, (0xC0 | encode));
 }
 
+void Assembler::vpcmpeqb(XMMRegister dst, XMMRegister src1, Address src2, int vector_len) {
```
Comment (sviswa7) [inline src/hotspot/cpu/x86/assembler_x86.cpp:4459]: InstructionMark missing in this instruction as well.

**Comment by asgibbons** [comment]
@sviswa7 Thanks for the good catches.  Fixed.

**Review by sviswa7 — APPROVED** [review]
Looks good to me.

**Comment by asgibbons** [comment]
Thanks!

/integrate

**Pull request closed without merging**